### PR TITLE
Update helm template -x to support children chart manifests in more cases

### PR DESCRIPTION
**Pull request by chancez**
The previous code validated the manifests exist within the file system,
which only works when a chart is unpacked, meaning children charts
stored as tgz's in the parent chart's `charts/` directory don't exist as
individual files in the filesystem, causing validation to file.

Instead, we validate against the paths within the chart itself rather
than the filesystem, handling both chart.tgz's and charts that are
unpacked as directories in the chart/ directory.

A simple example/use-case of where this previous failed, but does not after my change:

helm-operator is a dependency of chargeback-helm-operator, in my requirements.yaml. The `helm-operator.tgz` is obtained from `helm dep build/update`

```
$ tree chargeback-helm-operator
chargeback-helm-operator
├── Chart.yaml
├── charts
│   └── helm-operator-0.1.0.tgz
├── requirements.lock
├── requirements.yaml
└── values.yaml

1 directory, 5 files
```

And the helm-operator chart:
```
helm-operator
├── Chart.yaml
├── templates
│   ├── cr.yaml
│   ├── crd.yaml
│   ├── deployment.yaml
│   ├── role.yaml
│   ├── rolebinding.yaml
│   └── service-account.yaml
└── values.yaml

1 directory, 8 files
```

Given the above, the following would fail:
```
helm template charts/chargeback-helm-operator -x charts/helm-operator/templates/deployment.yaml
Error: could not resolve template path: stat /Users/chance/go/src/github.com/coreos-inc/kube-chargeback/charts/chargeback-helm-operator/charts/helm-operator/templates/deployment.yaml: no such file or directory
```

With this PRs changes the same command renders the `deployment.yaml` manifest correctly.

### CODE REVIEW

**Comment by chancez** [comment]
I took some liberties in refactoring/ordering the code since it was terribly difficult to understand before, as there was little comments and variable names weren't very descriptive/discernible.

**Review by thomastaylor312 — CHANGES_REQUESTED** [review]
@chancez Just one concern about how to handle non-existent paths. I tested out the code and if I give a bogus path, it returns 0 with no rendered manifests instead of an error. So it looks like something isn't catching things properly

**Comment by bacongobbler** [comment]
Perhaps some unit tests to cover these scenarios would be a good next approach :)

**Comment by chancez** [comment]
Oh..hm I had made some changes right before opening my PR that may have removed the error handling I did have. Let me revisit that.

**Comment by chancez** [comment]
@bacongobbler @thomastaylor312 I added some tests, and realized they actually still worked on master, then recalled my case was specifically with subcharts that are .tgzs, so I updated some of the testdata to support testing this use-case, and added a test that uses a subchart stored in a .tgz.

**Comment by chancez** [comment]
Also found another issue in how the tests were using subtests, so I fixed that. I also realized there were no tests that handle error cases, so I added a test that validates an error is thrown when passing a non-existent template to `-x`.

File: pkg/chartutil/testdata/mariner/templates/placeholder.tpl
Comment (thomastaylor312) [inline pkg/chartutil/testdata/mariner/templates/placeholder.tpl]: 💯

**Review by thomastaylor312 — APPROVED** [review]

**Comment by thomastaylor312** [comment]
@chancez Thanks for adding (and fixing) the unit tests on this!

**Comment by thomastaylor312** [comment]
Waiting on second review now that this is a size large

**Review by SlickNik — APPROVED** [review]
Reviewed this and did a quick smoke test as well. Thanks for the fix @chancez! LGTM 👍

**Pull request merged**
